File: ConsultaPoliza.Api/Controllers/PoliciesController.vb
Imports System.Diagnostics
Imports System.Globalization
Imports System.Net
Imports System.Text.RegularExpressions
Imports System.Threading
Imports System.Threading.Tasks
Imports System.Web.Http
Imports ConsultaPoliza.Api.Infrastructure
Imports ConsultaPoliza.Api.Models
Imports ConsultaPoliza.Api.Services
Imports Oracle.ManagedDataAccess.Client

Namespace ConsultaPoliza.Api.Controllers
    <RoutePrefix("api/polizas")>
    Public Class PoliciesController
        Inherits ApiController

        Private Shared ReadOnly PolicyNumberPattern As New Regex("^[0-9]+$", RegexOptions.Compiled)
        Private ReadOnly _policyRepository As IPolicyRepository

        Public Sub New()
            Me.New(ServiceRegistry.PolicyRepository)
        End Sub

        Friend Sub New(policyRepository As IPolicyRepository)
            _policyRepository = policyRepository
        End Sub

        <HttpGet>
        <Route("")>
        Public Async Function Search(
            ramo As Integer?,
            numeroPoliza As String,
            certificado As Integer?,
            fechaEfecto As DateTime?) As Task(Of IHttpActionResult)

            If Not ramo.HasValue OrElse ramo.Value <= 0 Then
                Return BadRequestMessage("Seleccione un ramo valido.")
            End If

            If Not IsValidPolicyNumber(numeroPoliza) Then
                Return BadRequestMessage("El numero de poliza es obligatorio y debe contener solamente numeros.")
            End If

            If Not certificado.HasValue OrElse certificado.Value < 0 Then
                Return BadRequestMessage("El certificado es obligatorio y no puede ser negativo.")
            End If

            If Not fechaEfecto.HasValue Then
                Return BadRequestMessage("La fecha de efecto es obligatoria.")
            End If

            Dim criteria = New PolicySearchCriteria(
                ramo.Value,
                Long.Parse(numeroPoliza.Trim(), CultureInfo.InvariantCulture),
                certificado.Value,
                fechaEfecto.Value.Date)

            Return Await FindPolicyAsync(
                Function() _policyRepository.SearchAsync(criteria, CancellationToken.None),
                numeroPoliza,
                ramo,
                certificado)
        End Function

        <HttpGet>
        <Route("{numeroPoliza}")>
        Public Async Function GetByNumber(numeroPoliza As String) As Task(Of IHttpActionResult)
            If Not IsValidPolicyNumber(numeroPoliza) Then
                Return BadRequestMessage("El numero de poliza es obligatorio y debe contener solamente numeros.")
            End If

            Return Await FindPolicyAsync(
                Function() _policyRepository.GetByNumberAsync(numeroPoliza.Trim(), CancellationToken.None),
                numeroPoliza,
                Nothing,
                Nothing)
        End Function

        Private Function BadRequestMessage(message As String) As IHttpActionResult
            Return Content(
                HttpStatusCode.BadRequest,
                New ErrorResponse With {
                    .Message = message
                })
        End Function

        Private Async Function FindPolicyAsync(
            findPolicy As Func(Of Task(Of PolicyResponse)),
            policyNumber As String,
            branchCode As Integer?,
            certificateNumber As Integer?) As Task(Of IHttpActionResult)

            Try
                Dim policy = Await findPolicy()
                If policy Is Nothing Then
                    Return Content(
                        HttpStatusCode.NotFound,
                        New ErrorResponse With {
                            .Message = "No se encontro una poliza con los criterios indicados."
                        })
                End If

                Return Ok(policy)
            Catch ex As OracleException When IsOracleConnectionError(ex)
                Trace.TraceError(
                    "Oracle connection failed while consulting policy {0}, branch {1}, certificate {2}: {3}",
                    policyNumber,
                    If(branchCode.HasValue, branchCode.Value.ToString(CultureInfo.InvariantCulture), ""),
                    If(certificateNumber.HasValue, certificateNumber.Value.ToString(CultureInfo.InvariantCulture), ""),
                    ex)

                Return Content(
                    HttpStatusCode.ServiceUnavailable,
                    New ErrorResponse With {
                        .Title = "Oracle no esta disponible.",
                        .Detail = "No se pudo conectar con Oracle. Verifique la red o VPN y que el servidor este disponible."
                    })
            Catch ex As Exception
                Trace.TraceError(
                    "Error consulting policy {0}, branch {1}, certificate {2}: {3}",
                    policyNumber,
                    If(branchCode.HasValue, branchCode.Value.ToString(CultureInfo.InvariantCulture), ""),
                    If(certificateNumber.HasValue, certificateNumber.Value.ToString(CultureInfo.InvariantCulture), ""),
                    ex)

                Return Content(
                    HttpStatusCode.InternalServerError,
                    New ErrorResponse With {
                        .Title = "No se pudo consultar la poliza.",
                        .Detail = "Ocurrio un error tecnico al consultar Oracle."
                    })
            End Try
        End Function

        Private Shared Function IsValidPolicyNumber(value As String) As Boolean
            If String.IsNullOrWhiteSpace(value) OrElse value.Length > 18 Then
                Return False
            End If

            Dim policyNumber As Long
            Return PolicyNumberPattern.IsMatch(value) AndAlso
                Long.TryParse(value, NumberStyles.None, CultureInfo.InvariantCulture, policyNumber) AndAlso
                policyNumber > 0
        End Function

        Private Shared Function IsOracleConnectionError(exception As OracleException) As Boolean
            If exception.Number = 50201 OrElse exception.Number = 50232 Then
                Return True
            End If

            For Each [error] As OracleError In exception.Errors
                If [error].Number = 50201 OrElse [error].Number = 50232 Then
                    Return True
                End If
            Next

            Return False
        End Function
    End Class
End Namespace

File: ConsultaPoliza.Api/Controllers/PoliciesController.cs
using ConsultaPoliza.Api.Models;
using ConsultaPoliza.Api.Services;
using Microsoft.AspNetCore.Mvc;
using Oracle.ManagedDataAccess.Client;
using System.Globalization;
using System.Text.RegularExpressions;

namespace ConsultaPoliza.Api.Controllers;

[ApiController]
[Route("api/polizas")]
public sealed partial class PoliciesController : ControllerBase
{
    private readonly IPolicyRepository _policyRepository;
    private readonly ILogger<PoliciesController> _logger;

    public PoliciesController(IPolicyRepository policyRepository, ILogger<PoliciesController> logger)
    {
        _policyRepository = policyRepository;
        _logger = logger;
    }

    [HttpGet]
    [ProducesResponseType(typeof(PolicyResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    [ProducesResponseType(StatusCodes.Status500InternalServerError)]
    public async Task<ActionResult<PolicyResponse>> Search(
        [FromQuery] int? ramo,
        [FromQuery] string? numeroPoliza,
        [FromQuery] int? certificado,
        [FromQuery] DateOnly? fechaEfecto,
        CancellationToken cancellationToken)
    {
        if (!ramo.HasValue || ramo.Value <= 0)
        {
            return BadRequest(new { message = "Seleccione un ramo valido." });
        }

        if (!IsValidPolicyNumber(numeroPoliza))
        {
            return BadRequest(new { message = "El numero de poliza es obligatorio y debe contener solamente numeros." });
        }

        if (!certificado.HasValue || certificado.Value < 0)
        {
            return BadRequest(new { message = "El certificado es obligatorio y no puede ser negativo." });
        }

        if (!fechaEfecto.HasValue)
        {
            return BadRequest(new { message = "La fecha de efecto es obligatoria." });
        }

        var criteria = new PolicySearchCriteria(
            RamoCodigo: ramo.Value,
            NumeroPoliza: long.Parse(numeroPoliza!.Trim(), CultureInfo.InvariantCulture),
            NumeroCertificado: certificado.Value,
            FechaEfecto: fechaEfecto.Value.ToDateTime(TimeOnly.MinValue));

        return await FindPolicyAsync(
            () => _policyRepository.SearchAsync(criteria, cancellationToken),
            numeroPoliza,
            ramo,
            certificado);
    }

    [HttpGet("{numeroPoliza}")]
    [ProducesResponseType(typeof(PolicyResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    [ProducesResponseType(StatusCodes.Status500InternalServerError)]
    public async Task<ActionResult<PolicyResponse>> GetByNumber(string numeroPoliza, CancellationToken cancellationToken)
    {
        if (!IsValidPolicyNumber(numeroPoliza))
        {
            return BadRequest(new { message = "El numero de poliza es obligatorio y debe contener solamente numeros." });
        }

        return await FindPolicyAsync(
            () => _policyRepository.GetByNumberAsync(numeroPoliza.Trim(), cancellationToken),
            numeroPoliza,
            branchCode: null,
            certificateNumber: null);
    }

    private async Task<ActionResult<PolicyResponse>> FindPolicyAsync(
        Func<Task<PolicyResponse?>> findPolicy,
        string policyNumber,
        int? branchCode,
        int? certificateNumber)
    {
        try
        {
            var policy = await findPolicy();
            if (policy is null)
            {
                return NotFound(new { message = "No se encontro una poliza con los criterios indicados." });
            }

            return Ok(policy);
        }
        catch (OracleException ex) when (IsOracleConnectionError(ex))
        {
            _logger.LogError(
                ex,
                "Oracle connection failed while consulting policy {PolicyNumber}, branch {BranchCode}, certificate {CertificateNumber}.",
                policyNumber,
                branchCode,
                certificateNumber);
            return Problem(
                title: "Oracle no esta disponible.",
                detail: "No se pudo conectar con Oracle. Verifique la red o VPN y que el servidor este disponible.",
                statusCode: StatusCodes.Status503ServiceUnavailable);
        }
        catch (Exception ex)
        {
            _logger.LogError(
                ex,
                "Error consulting policy {PolicyNumber}, branch {BranchCode}, certificate {CertificateNumber}.",
                policyNumber,
                branchCode,
                certificateNumber);
            return Problem(
                title: "No se pudo consultar la poliza.",
                detail: "Ocurrio un error tecnico al consultar Oracle.",
                statusCode: StatusCodes.Status500InternalServerError);
        }
    }

    private static bool IsValidPolicyNumber(string? value)
    {
        return !string.IsNullOrWhiteSpace(value)
            && value.Length <= 18
            && PolicyNumberRegex().IsMatch(value)
            && long.TryParse(value, NumberStyles.None, CultureInfo.InvariantCulture, out var policyNumber)
            && policyNumber > 0;
    }

    [GeneratedRegex("^[0-9]+$")]
    private static partial Regex PolicyNumberRegex();

    private static bool IsOracleConnectionError(OracleException exception)
    {
        return exception.Number is 50201 or 50232
            || exception.Errors.Cast<OracleError>().Any(error => error.Number is 50201 or 50232);
    }
}

/*
Cuando llega GET /api/polizas/203561, ASP.NET entra en GetByNumber.
Primero valida que 203561 sea numérico, positivo y razonable.
Después llama a _policyRepository.GetByNumberAsync, que es donde realmente se abre Oracle y se consulta la póliza.
Si no hay datos, responde 404.
Si hay datos, responde 200.
Si Oracle no está disponible, responde 503.
Si ocurre otro error técnico, responde 500.
El controlador no contiene SQL ni lógica de Oracle directa.
Su responsabilidad está bien separada: validar HTTP, coordinar el repositorio y traducir resultados/errores a respuestas HTTP.
*/
